[cxx-interop] Import parameterized public ctors of C++ foreign ref types as Swift Initializer

fahadnayyar opened the pull request:
Extends [PR #79986](https://github.com/swiftlang/swift/pull/79986) by adding support for calling parameterized C++ initializers from Swift. This patch synthesizes static factory methods corresponding to C++ parameterized constructors, allowing Swift to call them as Swift initializers (e.g., `init(_:)`, `init(_:_:)`, etc.).

rdar://148285251

fahadnayyar committed:
[cxx-interop] Support first non default ctor of c++ foreign reference types

fahadnayyar committed:
Added support for multiple ctors

fahadnayyar committed:
Added new tests for parameterized ctors

fahadnayyar commented:
**Note:** Default arguments are still not supported for both default and parameterized ctors

fahadnayyar commented:
@swift-ci please smoke test macOS

fahadnayyar commented:
@swift-ci please smoke test

Xazax-hun commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,167 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+      funcName += "_" + std::to_string(ctorParamCount) + "_params";
+    clang::IdentifierInfo *funcNameToSynthesize = &clangIdents.get(funcName);
+
+    llvm::SmallVector<clang::QualType, 4> paramTypes;
```
You might not need to create a new vector here. You could get the type of the ctor, which should be a `clang::FunctionProtoType`. And you could call `getParamTypes` which should return an `ArrayRef` to already allocated storage that has this information. This can shave off some allocations making the code more efficient (especially with ctors that have lots of parameters).

Xazax-hun commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,167 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+          clangCtx, synthesizedCxxMethodDecl, cxxRecordDeclLoc,
+          cxxRecordDeclLoc, origParam->getIdentifier(), origParam->getType(),
+          clangCtx.getTrivialTypeSourceInfo(origParam->getType()),
+          clang::SC_None, nullptr);
```
We are dropping the default argument here. Is that intentional? I thought you said you tested it and it works.

fahadnayyar replied:
Let's come back to supporting default arguments for ctors of foreign reference types (as well as value types) later: rdar://148575475

Xazax-hun commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,167 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+    synthesizedCxxMethodDecl->setParams(synthesizedParams);
+
+    if (!hasImmortalAttrs(cxxRecordDecl)) {
+      clang::SwiftAttrAttr *returnsRetainedAttrForSynthesizedCxxMethodDecl =
```
Nit: this variable name is really verbose and doe snot add any additional information to the code. I think you could just move the `Create` call inside the `addAttr` so now we don't need to pick a name here.

Xazax-hun commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,167 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+        clang::SwiftNameAttr::Create(clangCtx, swiftInitStr);
+    synthesizedCxxMethodDecl->addAttr(swiftNameAttr);
+
+    llvm::SmallVector<clang::Expr *, 8> ctorArgs;
```
Other small vectors had 4 as their inline buffer size, this one has 8. I think we should pick one and stick to that in the whole function unless we have a reason to diverge.

Xazax-hun replied:
This was mistakenly marked as resolved. I still see different values for the buffer sizes.

Xazax-hun commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,167 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+      auto *origParam = selectedCtorDecl->getParamDecl(i);
+      auto *param = clang::ParmVarDecl::Create(
+          clangCtx, synthesizedCxxMethodDecl, cxxRecordDeclLoc,
+          cxxRecordDeclLoc, origParam->getIdentifier(), origParam->getType(),
```
`origParam->getIdentifier()` can be a bit problematic. Consider the following case:
```
struct X {
  X(T);
};
```

Here, we specified the type for the argument but did not give it a name. In this case there is no identifier and we would not be able to refer to this parameter. Let's add a test case to test this scenario. Also add some code that generates a synthesized name in case the original parameter has no identifier.

Xazax-hun commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,167 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+
+    llvm::SmallVector<clang::ParmVarDecl *, 4> synthesizedParams;
+    for (unsigned int i = 0; i < ctorParamCount; ++i) {
+      auto *origParam = selectedCtorDecl->getParamDecl(i);
```
I think we might not want to create factories for constructors with variadic parameters, e.g.:
```
struct X {
  X(...) {}
};
```

Let's add a test case and make sure we don't to anything for these.

Xazax-hun commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,167 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+
+    llvm::SmallVector<clang::Expr *, 8> ctorArgs;
+    for (auto *param : synthesizedParams) {
+      ctorArgs.push_back(clang::DeclRefExpr::Create(
```
I think this is problematic for rvalue references. Consider:
```
struct X {
  X(T&& t);
};
```

We would synthesize something like:
```
X *_Nonnull synthesized(T&& t) {
  return new X(t);
}
```

The problem is that `t` in `X(t)` is an lvalue, so this does not compile giving me the following error with Clang:
```
<source>:7:14: error: no matching constructor for initialization of 'X'
    7 |   return new X(t);
      |              ^ ~
<source>:2:8: note: candidate constructor (the implicit copy constructor) not viable: cannot convert argument of incomplete type 'T' to 'const X' for 1st argument
    2 | struct X {
      |        ^
<source>:2:8: note: candidate constructor (the implicit move constructor) not viable: cannot convert argument of incomplete type 'T' to 'X' for 1st argument
    2 | struct X {
      |        ^
<source>:3:3: note: candidate constructor not viable: expects an rvalue for 1st argument
    3 |   X(T&& t);
      |   ^ ~~~~~
  ```
  
  To make this work, in case a parameter was an rvalue reference, we would need to insert an explicit cast like this:
```
X *_Nonnull synthesized(T&& t) {
  return new X(static_cast<T&&>(t));
}
```

Xazax-hun replied:
Let's add a test case with rvalue references.

Xazax-hun commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,167 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+          param->getType(), clang::VK_LValue));
+    }
+    llvm::SmallVector<clang::Expr *, 8> ctorArgsToAdd;
+    if (clangSema.CompleteConstructorCall(selectedCtorDecl, cxxRecordTy,
```
Could this `clangSema.CompleteConstructorCall` also introduce compile time errors similar to `FindAllocationFunctions`? If that is the case we also need to make sure this one would not fail the compilation similar to `FindAllocationFunctions`.

Xazax-hun commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,167 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+                                          ctorArgsToAdd))
+      continue;
+
+    clang::ExprResult synthesizedConstructExprResult =
```
Nit: The whole function is about synthesizing declarations and expressions. I think we can drop the "synthesized" prefix from these identifier names. It is clear from the context that we are creating them.

Xazax-hun commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,167 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+    assert(
+        !synthesizedNewExprResult.isInvalid() &&
+        "Unable to synthesize `new` expression for c++ foreign reference type");
+    clang::CXXNewExpr *synthesizedNewExpr =
```
Nit: you can use auto here. The type is already there in the cast expression, you don't need to repeat it twice in the same statement.

Xazax-hun commented on test/Interop/Cxx/class/Inputs/constructors.h:
```diff
@@ -307,15 +307,15 @@ __attribute__((swift_attr("release:Release10"))) DeletedCtor {
 struct __attribute__((swift_attr("import_reference")))
 __attribute__((swift_attr("retain:Retain11")))
-__attribute__((swift_attr("release:Release11"))) CtorWithDefaultArg {
+__attribute__((swift_attr("release:Release11"))) CtorWithDefaultArg { // expected-note {{'init(_:)' declared here}}
```
Where are these notes coming from? I think we should not emit them as the user never actually created these annotations. That being said, I am OK doing this in a follow-up PR.

fahadnayyar commented:
@swift-ci please smoke test macOS

fahadnayyar commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,167 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
-  }
-  if (!defaultCtorDecl)
-    return nullptr;
+  llvm::SmallVector<clang::CXXMethodDecl *, 4> synthesizedFactories;
```
Given that we are now iterating over multiple `ctorDecls`, it might be more logical to first check for operator `new` (which may be fewer in number than the constructors) before starting the iteration over the constructors. 

Additionally, since we're not reusing the `operatorNew` returned from `FindAllocationFunctions`, but instead calling `BuildCXXNew` (which internally may invoke `FindAllocationFunctions` again) to build the expression calling  operator `new`. 

What do you think?

j-hui replied:
@fahadnayyar I'm not sure I follow. Are you suggesting that we should call `clangSema.FindAllocationFunctions()` first because doing so might be more efficient? I don't know if we should assume whether `FindAllocationFunctions()` is more efficient, even if there are fewer `operator new`s.

Performance aside, I still think it makes more sense to make an extra pass through the ctors here at the beginning, to make it explicit that we won't synthesize any static factory methods if we can't find viable constructors, and that are giving up early on. (See also my comment below about explicitly returning an empty vec.)

Xazax-hun replied:
`FindAllocationFunctions` is not just iterating over the allocation functions. It is doing name lookup, overload resolution, traversing inheritance hierarchies and more. Also, we will need to figure out how to suppress the diagnostics from that call so this piece of code is likely to get more complex.

fahadnayyar replied:
Thanks so much for clarifying this suggestion, @j-hui!

fahadnayyar committed:
Addresses minor comments + refactoring

fahadnayyar commented:
@swift-ci please smoke test

j-hui commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,160 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
       operatorNew->getAccess() == clang::AS_private ||
       operatorNew->getAccess() == clang::AS_protected)
-    return nullptr;
+    return synthesizedFactories;
```
nit: perhaps change this to `return {};` or something to imply that we are intentionally returning an empty list because this is an exceptional case?

fahadnayyar committed:
Addressed some more comments

fahadnayyar commented:
@swift-ci please smoke test

fahadnayyar committed:
Default args in ctors (-) rvalue ref in ctors (+) identifierless param (+)

fahadnayyar commented:
@swift-ci please smoke test

Xazax-hun commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,192 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+        ctorDecl->getAccess() == clang::AS_protected ||
+        ctorDecl->isCopyOrMoveConstructor() || ctorDecl->isVariadic())
+      continue;
+    bool hasDefaultArg = false;
```
You don't need a loop to check if a function has a default argument. It is enough to only look at the last parameter (if there is at least one). This is because in C++ if a parameter has default argument, all the following ones also must have default arguments. So you can get rid of this loop.

Xazax-hun commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,192 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+        break;
+      }
     }
+    // Todo: Add support for default args in ctors for C++ foreign reference
```
Nit: we usually capitalize `TODO` and terminate full sentences with periods in comments.

Xazax-hun commented on test/Interop/Cxx/class/Inputs/constructors.h:
```diff
@@ -337,6 +337,34 @@ __attribute__((swift_attr("release:Release14"))) ParameterizedCtor {
+struct __attribute__((swift_attr("import_reference")))
+__attribute__((swift_attr("retain:Retain18")))
+__attribute__((swift_attr("release:Release18"))) NoIdentifierInCtorParam {
+public:
```
Would be nice to have some tests with constructors of user defined types that have user defined copy/move ctors.

Xazax-hun commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,192 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+
+    std::string swiftInitStr = "init(";
+    for (unsigned i = 0; i < ctorParamCount; ++i) {
+      swiftInitStr += "_:";
```
Actually, in case the corresponding parameter is an rvalue reference, this should be `"consuming:"`.

fahadnayyar committed:
Addressed reaminging comments

fahadnayyar commented:
@swift-ci please smoke test

fahadnayyar commented:
Enabling auto merge after addressing all the comments on this patch. I can think of only 2 things right now which can be addressed in a follow up patch:

- Default args in C++ frt ctors
- Better Tests for Copy, move and other kinds of ctors of C++ FRTs

Xazax-hun commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -2534,128 +2534,204 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+        "Unable to synthesize `new` expression for c++ foreign reference type");
+    auto *synthNewExpr = cast<clang::CXXNewExpr>(synthNewExprResult.get());
+
+    clang::ReturnStmt *synthRetStmt = clang::ReturnStmt::Create(
```
Nit: I wonder if we want to use `Sema::BuildReturnStmt` here, and add the `SFNAETrap` to all of the sema calls (could be the same trap for all the `Build*` calls).

Xazax-hun reviewed (approved):
This looks great! Thanks for working on this! I have one last comment inline but otherwise everything looks good to me.

fahadnayyar committed:
macOS CI issue fix + SFINAETrap protection on all clangSema APIs

fahadnayyar commented:
@swift-ci please smoke test

fahadnayyar commented:
Enabling auto merge after addressing all the comments on this patch. I can think of only these minor things which can be addressed in follow up patches later:

- Default args in C++ frt ctors. They are also not supported for ctors of Value types.
- Better Tests for Copy, move and other kinds of ctors of C++ FRTs so that the tests make sense on the Swift side.
- A lit test involving <string> has been removed for now to make macOS swift ci happy because of some unrelated issue: rdar://148791414. But I verified locally that the removed test also works well.
